processor/batchprocessor/batch_processor.go

```diff
@@ -98,120 +199,188 @@ func (bp *batchProcessor) Capabilities() consumer.Capabilities {
+				b.sendItems(triggerTimeout)
 			}
-			bp.resetTimer()
+			b.resetTimer()
 		}
 	}
 }
 
-func (bp *batchProcessor) processItem(item any) {
-	bp.batch.add(item)
+func (b *batcher) processItem(item any) {
+	b.batch.add(item)
 	sent := false
-	for bp.batch.itemCount() >= bp.sendBatchSize {
+	for b.batch.itemCount() >= b.processor.sendBatchSize {
 		sent = true
-		bp.sendItems(triggerBatchSize)
+		b.sendItems(triggerBatchSize)
 	}
 
 	if sent {
-		bp.stopTimer()
-		bp.resetTimer()
+		b.stopTimer()
+		b.resetTimer()
 	}
 }
 
-func (bp *batchProcessor) stopTimer() {
-	if !bp.timer.Stop() {
-		<-bp.timer.C
+func (b *batcher) stopTimer() {
+	if !b.timer.Stop() {
+		<-b.timer.C
 	}
 }
 
-func (bp *batchProcessor) resetTimer() {
-	bp.timer.Reset(bp.timeout)
+func (b *batcher) resetTimer() {
+	b.timer.Reset(b.processor.timeout)
 }
 
-func (bp *batchProcessor) sendItems(trigger trigger) {
-	sent, bytes, err := bp.batch.export(bp.exportCtx, bp.sendBatchMaxSize, bp.telemetry.detailed)
+func (b *batcher) sendItems(trigger trigger) {
+	sent, bytes, err := b.batch.export(b.exportCtx, b.processor.sendBatchMaxSize, b.processor.telemetry.detailed)
 	if err != nil {
-		bp.logger.Warn("Sender failed", zap.Error(err))
+		b.processor.logger.Warn("Sender failed", zap.Error(err))
 	} else {
-		bp.telemetry.record(trigger, int64(sent), int64(bytes))
+		b.processor.telemetry.record(trigger, int64(sent), int64(bytes))
+	}
+}
+
+func (bp *batchProcessor) findBatcher(ctx context.Context) (*batcher, error) {
+	if bp.singleton != nil {
+		return bp.singleton, nil
+	}
+
+	// Get each metadata key value, form the corresponding
+	// attribute set for use as a map lookup key.
+	info := client.FromContext(ctx)
+	md := map[string][]string{}
+	var attrs []attributeKeyValue
+	for _, k := range bp.metadataKeys {
+		// Lookup the value in the incoming metadata, copy it
+		// into the outgoing metadata, and create a unique
+		// value for the attributeSet.
+		vs := info.Metadata.Get(k)
+		md[k] = vs
+		if len(vs) == 1 {
+			attrs = append(attrs, attributeString(k, vs[0]))
+		} else {
+			attrs = append(attrs, attributeStringSlice(k, vs))
+		}
+	}
+	aset := newAttributeSet(attrs...)
+
+	bp.lock.Lock()
+	defer bp.lock.Unlock()
+
+	b, ok := bp.batchers[aset]
+	if ok {
+		return b, nil
+	}
+
+	if limit := bp.metadataLimit; limit != 0 && len(bp.batchers) >= int(limit) {
+		return nil, errTooManyBatchers
 	}
+
+	// aset.ToSlice() returns the sorted, deduplicated,
+	// and name-downcased list of attributes.
+	b = bp.newBatcher(md)
+	bp.batchers[aset] = b
+	return b, nil
+}
+
+func (bp *batchProcessor) currentMetadataCardinality() int {
+	bp.lock.Lock()
+	defer bp.lock.Unlock()
+	return len(bp.batchers)
 }
 
 // ConsumeTraces implements TracesProcessor
-func (bp *batchProcessor) ConsumeTraces(_ context.Context, td ptrace.Traces) error {
-	bp.newItem <- td
+func (bp *batchProcessor) ConsumeTraces(ctx context.Context, td ptrace.Traces) error {
+	b, err := bp.findBatcher(ctx)
```

Comment:
I feel like there's room for a small improvement here: given that the decision to use either a singleton batcher or individual batchers is known when the processor is built, perhaps each could have its own implementation of `findBatcher` and this would be set as part of the processor? The hot path here would then just call the right implementation. 

In the current implementation, it saves only a conditional check, which might not justify the improvement though.

Reply:
Done.